### PR TITLE
Alias management attempt nr. 2

### PR DESCRIPTION
**Pull request by DevManu-de**
I rewrote the `alias.c`.

- Now you can declare multiple aliases at once
- Decreased Code size
- Removed unneeded `xmalloc`, `xcalloc` and `xrealloc`

### CODE REVIEW

File: src/builtins/alias.c [hunk 1]
```diff
@@ -43,33 +43,40 @@ int builtin_alias(int argc, char **argv, psh_state *state)
...
+        /* If  alias is already set print error otherwise add it */
         if (check_for_alias(state->alias_table, alias) == NULL)
         {
             add_alias(state->alias_table, alias, value);
-            puts(check_for_alias(state->alias_table, alias));
 
         } else
         {
-            puts(check_for_alias(state->alias_table, alias));
             OUT2E("The alias: %s already exists with %s\n", alias, check_for_alias(state->alias_table, alias));
```
Comment (DevManu-de) [inline src/builtins/alias.c]: I know there is a bug where `check_for_alias(state->alias_table, alias)` is not printed when it should.
It prints nothing, not even some trash characters.
I am working on this right now.

**Comment by DevManu-de** [comment]
In this latest commit I added the ability to display all aliases with `alias` like bash's `alias`.

**Comment by DevManu-de** [comment]
Does a function already exist that replaces a piece of a command with another piece or do I have to write this from scratch.

**Comment by myzhang1029** [comment]
> Does a function already exist that replaces a piece of a command with another piece or do I have to write this from scratch.

No, such a function doesn't exist. Please note that you should operate on `buffer` in `main.c` instead of the `struct _psh_command` in `run.c`, otherwise aliases with spaces would be treated as one single command name.

**Comment by DevManu-de** [comment]
Should I move `expand_alias()` to another file or is it fine if I keep it in the builtin `alias.c`?

**Comment by DevManu-de** [comment]
Missing features:
- [x] unalias -a flag

Known bugs:
- [x] exit command does nothing after wrong input was given
- [x] exit command produces segmentation fault
Both were fixed after a merge of your latest commits

- [x] Documentation

**Comment by DevManu-de** [comment]
Everything should work now.
When you test it and find a bug just tell me.
Its impossible to check every possibility.

Looking forward to your reviews

**Comment by DevManu-de** [comment]
I have found a bug with unalias.
Currently unalias is stuck in an infinite loop.
Fix: add a `+` on line 93 in `alias.c` to fix the infinite loop.
`for (i = 1; i < argc; + i)` -> `for (i = 1; i < argc; ++i)`

Furthermore I want to know why it was necessary for you to rewrite the entire `alias.c` file.

**Comment by myzhang1029** [comment]
> I have found a bug with unalias.
> Currently unalias is stuck in an infinite loop.
> Fix: add a `+` on line 93 in `alias.c` to fix the infinite loop.
> `for (i = 1; i < argc; + i)` -> `for (i = 1; i < argc; ++i)`

Fixed.

> Furthermore I want to know why it was necessary for you to rewrite the entire `alias.c` file.

I didn't.
I only changed the builtins for two reasons:
1. for conformance with bash: bash allowed updating aliases but your code reported an error. bash printed the current value but your code reported another error.
2. to remove the global array because if this project will ever expose a C interface, that would be not thread-safe. Also, if we have such an index, it's completely removing the benefit of a hash table.

For the expansion function, one is again bash conformance, another is to use the C library functions more efficiently (such as `strstr`), there were also memory leaks caused by the `strdup`s on line 205, 110 and 111 which I didn't want to go over the logic to fix.

**Comment by DevManu-de** [comment]
Ok I get why it was necessary to change the `alias.c`
Maybe its the best for both of us that I stop making pull request because all you have to do is to fix my mistakes and this only stops you from making progress in this project. So dont expect a pull request from me in the near future maybe later.

**Comment by DevManu-de** [comment]
Could you please explain me why my `strdup` on line 111 is a memory leak since you do the same on line 66 with `psh_strdup` and I wasn't able to find a significant difference in `psh_strdup` and `strdup`.
Sorry if I annoy you with my questions but I want to lean and understand why.
But for the others I agree that I made some mistakes.

**Comment by myzhang1029** [comment]
> Could you please explain me why my `strdup` on line 111 is a memory leak since you do the same on line 66 with `psh_strdup` and I wasn't able to find a significant difference in `psh_strdup` and `strdup`.
> Sorry if I annoy you with my questions but I want to lean and understand why.
Yes of course. Don't worry, if in doubt, just ask.

It is not about the difference in that two functions. Have you ever noticed the fourth argument to `psh_hash_add_chk`? You passed 0 to it on [line 37](https://github.com/myzhang1029/psh/blob/f29d52363d714d4dc5a91e3d750337d4992382be/src/builtins/alias.c#L37), but in the [documentation for that function](https://github.com/myzhang1029/psh/blob/f29d52363d714d4dc5a91e3d750337d4992382be/include/libpsh/hash.h#L112), you can see that 0 indicates that the value will not be deallocated upon the deallocation of this hash table. I don't think you can learn  much from this specific case, if I have to come up with one, maybe it is, always read the documentation and understand every argument before using an interface by someone else.

Do ask if you have other questions. These days I'm having much classwork so I'm sorry if my language is impatient sometimes.

**Pull request merged**